### PR TITLE
cpu/atmega2560: fixes the configuration of pullup resistors in gpio.c

### PR DESCRIPTION
**Pull request by danielamkaer**
> "If PORTxn is written logic one when the pin is configured as an input pin, the pull-up resistor is activated. To switch the pull-up resistor off, PORTxn has to be written logic zero or the pin has to be configured as an output pin. The port pins are tri-stated when reset condition becomes active, even if no clocks are running."

[[datasheet](http://atmel.com/Images/Atmel-2549-8-bit-AVR-Microcontroller-ATmega640-1280-1281-2560-2561_datasheet.pdf) p. 68]

### CODE REVIEW

File: cpu/atmega2560/periph/gpio.c [hunk 1]
```diff
@@ -121,8 +139,6 @@ int gpio_init_int(gpio_t pin, gpio_pp_t pullup, gpio_flank_t flank,
         return -1;
     }
 
-    gpio_set(pin);
```
Comment (thomaseichinger) [inline cpu/atmega2560/periph/gpio.c]: Is this change intended?

Comment (danielamkaer) [inline cpu/atmega2560/periph/gpio.c]: Yes. gpio_set will set the bit in PORTx aswell. enabling the pullup no matter  what

**Comment by PeterKietzmann** [comment]
@ReneHerthel, want to have a look at it and try on hardware?

**Comment by ReneHerthel** [comment]
@PeterKietzmann Yep :)

File: cpu/atmega2560/periph/gpio.c [hunk 2]
```diff
@@ -102,6 +102,24 @@ int gpio_init(gpio_t pin, gpio_dir_t dir, gpio_pp_t pullup)
     else {
         _SFR_MEM8(_ddr_addr(pin)) &= ~(1 << _pin_num(pin));
         res = bit_is_clear(_SFR_MEM8(_ddr_addr(pin)), _pin_num(pin));
+
+        if (res == 0) {
```
Comment (ReneHerthel) [inline cpu/atmega2560/periph/gpio.c]: @danielamkaer Are you sure with this?

Macro bit_is_clear:

```Javascript
#define bit_is_clear(sfr,bit) \
	(!(_SFR_BYTE(sfr) & _BV(bit))) 
```

Test whether bit bit in IO register sfr is clear. This will return **non-zero** if the bit is **clear**, and a **0** if the bit is **set**. 

And we want to clear the bit, or am I wrong?

File: cpu/atmega2560/periph/gpio.c [hunk 3]
```diff
@@ -102,6 +102,24 @@ int gpio_init(gpio_t pin, gpio_dir_t dir, gpio_pp_t pullup)
...
+        }
+
+        switch (pullup) {
+            case GPIO_NOPULL:
+                _SFR_MEM8(_port_addr(pin)) &= ~(1 << _pin_num(pin));
+                res = bit_is_clear(_SFR_MEM8(_port_addr(pin)), _pin_num(pin));
+                break;
+            case GPIO_PULLUP:
+                _SFR_MEM8(_port_addr(pin)) |= (1 << _pin_num(pin));
+                res = bit_is_clear(_SFR_MEM8(_port_addr(pin)), _pin_num(pin));
```
Comment (ReneHerthel) [inline cpu/atmega2560/periph/gpio.c]: @danielamkaer shouldn't be placed here `bit_is_set`?

File: cpu/atmega2560/periph/gpio.c (hunk 2 above)
Comment (danielamkaer) [inline cpu/atmega2560/periph/gpio.c]: If it's zero, return -1 in error.

File: cpu/atmega2560/periph/gpio.c (hunk 3 above)
Comment (danielamkaer) [inline cpu/atmega2560/periph/gpio.c]: Yes, I will change that.

File: cpu/atmega2560/periph/gpio.c (hunk 2 above)
Comment (ReneHerthel) [inline cpu/atmega2560/periph/gpio.c]: @danielamkaer Yes, this is what the if statement says, but the bit_is_clear macro returns a value which is **not 0** when the bit is **clear**.
The if statement checks if the return value is **equal 0** and that says, the bit is **set**.

>The DDxn bit in the DDRx Register selects the direction of this pin. If DDxn is written logic one, Pxn is configured
as an output pin. **If DDxn is written logic zero, Pxn is configured as an input pin.**

The bit must be cleared for input pins.
You ask if the bit is set.

BTW: I can't initialize input pins with this, it returns always -1

Comment (danielamkaer) [inline cpu/atmega2560/periph/gpio.c]: "The if statement checks if the return value is equal 0 and that says, the bit is set."
Yes, if the bit is still **set** after we **cleared** it, it is clearly an error and we should return -1 *(and not bother with the pull up configuration)*.

Can you show me the code your calling the init with, as it works for me.

Comment (ReneHerthel) [inline cpu/atmega2560/periph/gpio.c]: Ah, yes I see. I'm sorry this macros confused me a little :-).
I test it with RIOT/tests/periph_gpio, maybe the test is buggy. I'll look at it later.

Comment (ReneHerthel) [inline cpu/atmega2560/periph/gpio.c]: @danielamkaer The problem was the wrong macro in the lower lines, at the time I tested it I tried something like `init_in PORTA 4 1` and it had returned -1 :)

But now it should work :)

**Comment by ReneHerthel** [comment]
@PeterKietzmann @danielamkaer So it works on my hardware.

**Comment by PeterKietzmann** [comment]
ACK and go
